### PR TITLE
Borrow api module from containerd (instead of vendoring it)

### PR DESCRIPTION
**Pull request by dims**
We only need the `api` folder from https://github.com/containerd/containerd project. So we have a script that fetches a zip for a specific version (latest as of now) and sticks it into a `third_party/containerd` directory. We can run this script as and when needed and check in the code. We do this until https://github.com/containerd/containerd/issues/7231 is fixed on containerd-side. 

Note that we patch the imports as well as we are creating a new directory/package structure here as well.

### CODE REVIEW

**Comment by dims** [comment]
experiment was ✅ 

Looks like the presubmit(s) are broken because of golang changes. Let's formalize this next week.

**Comment by dims** [comment]
related to https://github.com/containerd/containerd/issues/7231

**Comment by akhilerm** [comment]
@dims Is this a workaround/hack that we are going to follow for this release?

**Comment by dims** [comment]
@akhilerm yes, that would be my proposal.

File: go.mod [hunk 1]
```diff
@@ -7,23 +7,22 @@ require (
 	github.com/Microsoft/go-winio v0.4.15
 	github.com/aws/aws-sdk-go v1.35.24
 	github.com/blang/semver v3.5.1+incompatible
-	github.com/containerd/containerd v1.4.12
 	github.com/containerd/ttrpc v1.0.2
 	github.com/containerd/typeurl v1.0.2
-	github.com/docker/distribution v2.8.0+incompatible // indirect
-	github.com/docker/docker v20.10.12+incompatible
+	github.com/docker/distribution v2.8.1+incompatible // indirect
+	github.com/docker/docker v20.10.17+incompatible
```
Comment (bobbypage) [inline go.mod:13]: Do we have to change this? k/k is pinned on v20.10.12 - https://github.com/kubernetes/kubernetes/blob/6aa5efb/go.mod#L342

File: go.mod [hunk 2]
```diff
@@ -7,23 +7,22 @@ require (
...
 	github.com/docker/go-connections v0.4.0
 	github.com/docker/go-units v0.4.0
 	github.com/euank/go-kmsg-parser v2.0.0+incompatible
+	github.com/gogo/googleapis v1.4.1
```
Comment (bobbypage) [inline go.mod:17]: Why do we need to take on a new dependency by removing containerd?

**Comment by bobbypage** [comment]
Hmm, I'm hitting some issues running the update script myself:

```
$ gh pr checkout https://github.com/google/cadvisor/pull/3145
porterdavid@porterdavid: ~/github/cadvisor on borrow-api-module-from-containerd [!?$]
$ rm -rf third_party/

porterdavid@porterdavid: ~/github/cadvisor on borrow-api-module-from-containerd [!?$]
$ build/update-containerd-api.sh
~/github/cadvisor/third_party/containerd ~/github/cadvisor
tar: Pattern matching characters used in file names
tar: Use --wildcards to enable pattern matching, or --no-wildcards to suppress this warning
tar: */api: Not found in archive
tar: Pattern matching characters used in file names
tar: Use --wildcards to enable pattern matching, or --no-wildcards to suppress this warning
tar: */LICENSE: Not found in archive
tar: Pattern matching characters used in file names
tar: Use --wildcards to enable pattern matching, or --no-wildcards to suppress this warning
tar: */NOTICE: Not found in archive
tar: Exiting with failure status due to previous errors
```

**Comment by bobbypage** [comment]
Thank you Dims for working on this, I think this will be a great workaround until containerd will have a proper API go module.

File: go.mod (hunk 1 above)
Comment (dims) [reply go.mod:13]: yes, we need this, the older version of docker/distribution + docker/docker will drag containerd back in.

File: go.mod (hunk 2 above)
Comment (dims) [reply go.mod:17]: See above ^ we really need to update docker/docker and docker/distribution, so that drags this in.

**Comment by dims** [comment]
@bobbypage what version of `tar` do you have? if this is on a macos box, it will usually have a really old tar right?

**Comment by bobbypage** [comment]
> @bobbypage what version of `tar` do you have? if this is on a macos box, it will usually have a really old tar right?

This on my linux box ~ Debian Testing

```
$ tar --version
tar (GNU tar) 1.34
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
```

**Comment by dims** [comment]
fixed that wildcards problem @bobbypage (looks like i did this work on my macos which had the older version of tar)

**Comment by bobbypage** [comment]
LGTM!

File: go.mod (hunk 2 above)
Comment (liggitt) [reply go.mod:17]: hmm... do we have a good sense for the stability/maintainedness of this new gogo repo? gogo/protobuf has me wary

Comment (dims) [reply go.mod:17]: @liggitt will dig into this for sure

Comment (akhilerm) [reply go.mod:17]: `gogo/googleapis` is also being used in the generated [introspection.pb.go](https://github.com/google/cadvisor/blob/master/third_party/containerd/api/services/introspection/v1/introspection.pb.go#L10)

Comment (dims) [reply go.mod:17]: @akhilerm exactly!

@liggitt the `main` branch of containerd/containerd no longer has this. It was removed in https://github.com/containerd/containerd/pull/6841

Here's how it looks now:
https://github.com/containerd/containerd/blob/main/api/services/introspection/v1/introspection.pb.go

So when containerd 1.7.x series starts we can switch to it and the dependency will get cleaned up.

**Comment by brandond** [comment]
FYI, this has completely broken projects like K3s that try to link to both Containerd and Kubelet code from the same project.

```
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.Container
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.Container.ExtensionsEntry
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.Container.LabelsEntry
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.Container.Runtime
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.GetContainerRequest
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.GetContainerResponse
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.ListContainersRequest
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.ListContainersResponse
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.CreateContainerRequest
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.CreateContainerResponse
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.UpdateContainerRequest
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.UpdateContainerResponse
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.DeleteContainerRequest
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.services.containers.v1.ListContainerMessage
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.types.Descriptor
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.types.Descriptor.AnnotationsEntry
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.types.Metric
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.types.Mount
2022/11/29 23:49:43 proto: duplicate proto type registered: containerd.types.Platform
panic: proto: duplicate enum registered: containerd.v1.types.Status

goroutine 1 [running]:
github.com/gogo/protobuf/proto.RegisterEnum(...)
	/go/pkg/mod/github.com/gogo/protobuf@v1.3.2/proto/properties.go:520
github.com/google/cadvisor/third_party/containerd/api/types/task.init.0()
	/go/pkg/mod/github.com/google/cadvisor@v0.46.0/third_party/containerd/api/types/task/task.pb.go:161 +0x1a5
	
brandond@dev01:~/go/src/github.com/k3s-io/k3s$ go mod why github.com/google/cadvisor/third_party/containerd/api/types/task
# github.com/google/cadvisor/third_party/containerd/api/types/task
github.com/k3s-io/k3s/pkg/daemons/executor
k8s.io/kubernetes/cmd/kubelet/app
k8s.io/kubernetes/cmd/kubelet/app/options
github.com/google/cadvisor/container/containerd
github.com/google/cadvisor/third_party/containerd/api/types/task

brandond@dev01:~/go/src/github.com/k3s-io/k3s$ go mod why github.com/containerd/containerd/api/types
# github.com/containerd/containerd/api/types
github.com/k3s-io/k3s/pkg/agent/containerd
github.com/containerd/containerd
github.com/containerd/containerd/api/types
```

**Comment by dims** [comment]
@brandond just to make sure whoever is looking at this. The Kubernetes team does not support the use case of kubelet code imported into other projects/products as mentioned in our README - https://github.com/kubernetes/kubernetes/#to-start-using-k8s pasting the snippet here:

```
To use Kubernetes code as a library in other applications, see the [list of published components](https://git.k8s.io/kubernetes/staging/README.md). Use of the k8s.io/kubernetes module or k8s.io/kubernetes/... packages as libraries is not supported.
```

**Comment by brandond** [comment]
Yes, we're aware that we are doing naughty things by mashing everything all together and mixing it up with a spoon. It mostly works, except when it doesn't.

**Pull request merged**